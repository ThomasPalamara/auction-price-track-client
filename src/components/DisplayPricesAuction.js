import React from "react";
import { Table } from 'antd';

const columns = [{
    title: 'Prix à l\'unité',
    dataIndex: 'price',
    key: 'price',
    render: (text) => (
        <div>
            <span>{text.toString().slice(0, -4)} <i className="money-icon gold"/> </span>
            <span>{text.toString().slice(-4, -2)} <i className="money-icon silver"/> </span>
            <span>{text.toString().slice(-2)} <i className="money-icon copper"/> </span>
        </div>
    ),
}, {
    title: 'Quantité',
    dataIndex: 'quantity',
    key: 'quantity',
}];


class DisplayPricesTable extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        // const auctions = this.props.auctions[0].prices.sort(function(a,b) {return (a.price > b.price) ? 1 : ((b.price > a.price) ? -1 : 0);} );
        console.log(this.props.auctions[0].prices);
        return (
            <Table
                columns={columns}
                dataSource={
                    this.props.auctions[0].prices.map((element, i) => {
                        return (
                            {
                                key: i,
                                price: element.price,
                                quantity: element.quantity,
                            }
                        );
                    })
                }
            />
        );
    }
}

export default DisplayPricesTable;


// <table>
// {props.auctions[0].prices.map((x) => {
//     return (
//         <tr>
//             <td>
//                 <span>{x.price.toString().slice(0, -4)} PO </span>
//                 <span>{x.price.toString().slice(-4, -2)} PA </span>
//                 <span>{x.price.toString().slice(-2)} PC </span>
//             </td>
//             <td>{x.quantity}</td>
//         </tr>
//     )
// }
// )}
// </table>