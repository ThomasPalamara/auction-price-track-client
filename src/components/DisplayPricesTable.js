import React from "react";
import { connect } from 'react-redux';
import DisplayPricesTable from './DisplayPricesTable';
import { InputNumber, Table } from 'antd';
import PriceCoinDisplay from "./PriceCoinDisplay";


class DisplayPrices extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
        }
    }


    getPriceForQuantity = (auctions, need) => {
        let qteactu = 0;
        let result = 0;
        auctions.prices.map((element) => {
            if (need > qteactu) {
                let qte = Math.min(need - qteactu, element.quantity);
                result += qte * element.price;
                qteactu += qte;
            }
        });
        return result;
    }

    render() {
        return (
            <div>
                <p>Prix de vos ventes</p>
                <Table
                    columns={[{
                        title: 'Nom',
                        dataIndex: 'name',
                        key: 'name',
                        render: (text,record) => (
                            <a
                                data-wowhead={`item=${record.key}`}
                                href="#"
                            >{text}</a>
                        ),
                    }, {
                        title: 'Prix total',
                        dataIndex: 'totalPrice',
                        key: 'totalPrice',
                        render: (text, record) => {
                            if (record.itemExists) {
                                return (<PriceCoinDisplay price={text}/>)
                            } else {
                                return <p>{record.text}</p>
                            }
                        },
                    }, {
                        title: 'Action',
                        key: 'action',
                        render: (text, record) => (
                            <span>
                                <InputNumber min={0} max={record.quantityMax} defaultValue={1} disabled={!record.itemExists}
                                    onChange={(value) => {
                                        this.props.dispatch(this.props.editQuantity({ id: record.key, quantity: value }))
                                    }} />
                            </span>
                        ),
                    }, {
                        title: 'Details des encheres',
                        key: 'auctionsDetails',
                        render: (text, record) => {
                            if (record.itemExists) {
                                // return <DisplayPricesTable auctions={record.auctions} /> 
                            } else {
                                return <p>!</p>
                            }
                        }
                    }]}
                    /*---- DATASOURCE ----*/
                    dataSource={
                        this.props.items.map((element, i) => {
                            const itemElem = this.props.auctionHouse.find(elem => elem.item_id === element.id.toString()); //object of itemModel from auctionHouse
                            if (!itemElem) {
                                return (
                                    {
                                        key: element.id,
                                        name: element.name[this.props.itemLanguage],
                                        auctions: null,
                                        totalPrice: "Aucune enchere pour cet objet",
                                        quantityMax: 0,
                                        itemExists: false
                                    }
                                )
                            } else {
                                return (
                                    {
                                        key: element.id,
                                        name: element.name[this.props.itemLanguage],
                                        auctions: itemElem.auctions[0],
                                        totalPrice: this.getPriceForQuantity(itemElem.auctions[0], element.quantity),
                                        quantityMax: itemElem.auctions[0].total_quantity,
                                        itemExists: true
                                    }
                                )
                            }

                        })
                    }
                />
            </div>
        )
    }
};

const mapStatetoProps = state => {

    return {
        product: state.product,
        ingredient: state.ingredient,
        itemLanguage: state.itemLanguage
    }
};

export default connect(mapStatetoProps)(DisplayPrices);

//Tableau de toutes les auctions pour chaque item
// {this.props.product &&
//     this.props.product.map((item) => {
//         const itemElem = this.props.auctionHouse.find(elem => elem.item_id === item.id.toString())
//         const totalPrice = item.quantity
//         return (
//             <div>
//                 <p>{item.name}</p>
//                 <InputNumber min={0} defaultValue={1}
//                     onChange={(value) => {
//                         this.props.dispatch(editQuantityProduct({ id: item.id, quantity: value }))
//                     }} />
//             </div>
//         )
//     })}