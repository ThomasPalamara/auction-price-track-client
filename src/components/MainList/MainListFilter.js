/* eslint-disable react/jsx-boolean-value */
/* Option setting a value at true is necessary */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Radio, Select, Input, Row, Col } from 'antd';
import capitalize from 'functions/capitalize';
import { setTextFilter, setProfessionFilter, setisCustomFilter } from '../../actions/filters';

const { Option } = Select;

// TODO Redo the whole text filter so it works with craft and recipe items
// TODO Decide if each filters should get its own component.
// TODO Found a solution for the scroll bar appearing while switching type of list on mobile.

// All filters values are sent to the store. The selector then filters the list in the MainList component.

const MainListFilter = ({ professions, filters, dispatch }) => (
  <React.Fragment>
    <Row className="filters__container" type="flex" justify="space-between" align="bottom">
      <Col order="1" className="filterGroup">
        <p className="filterGroup__label" htmlFor="professionFilter"> Type of list :</p>
        <Radio.Group
          value={filters.isCustom}
          onChange={e => dispatch(setisCustomFilter(e.target.value))}
        >
          <Radio.Button value={false}>Recipes</Radio.Button>
          <Radio.Button value={true}>Reagents</Radio.Button>
        </Radio.Group>
      </Col>

      <Col order="2" className="filterGroup">
        <p className="filterGroup__label" htmlFor="professionFilter">Profession :</p>
        <Select id="professionFilter" defaultValue="all" style={{ width: 120 }} onChange={value => dispatch(setProfessionFilter(value))}>
          <Option value="all">All</Option>
          {
            professions.map(profession => (
              <Option key={profession.toLowerCase()} value={profession.toLowerCase()}>
                {capitalize(profession)}
              </Option>
            ))
          }
        </Select>
      </Col>
      <Col order="3" className="filterGroup">
        <p className="filterGroup__label" htmlFor="textFilter">Search item name :</p>
        <Input
          id="textFilter"
          value={filters.text}
          style={{ width: 172 }}
          placeholder="Filter items"
          onChange={e => (dispatch(setTextFilter(e.target.value)))}
        />
      </Col>
      <Col order="4" />
    </Row>
  </React.Fragment>
);

MainListFilter.propTypes = {
  professions: PropTypes.arrayOf(PropTypes.string).isRequired,
  filters: PropTypes.shape({
    isCustom: PropTypes.bool,
    profession: PropTypes.string,
    text: PropTypes.string,
  }).isRequired,
  dispatch: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  const array = [];
  // Getting all the professions names from the list of recipes
  state.recipes.recipesList.map(recipe => recipe.professions.map(profession => (array.indexOf(profession) === -1 ? array.push(profession) : '')));

  return {
    filters: state.filters,
    professions: array,
  };
};

export default connect(mapStateToProps)(MainListFilter);
