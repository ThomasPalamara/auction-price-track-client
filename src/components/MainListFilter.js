import React from "react";
import { connect } from "react-redux";
import { Radio, Select,  } from 'antd';
import { setTextFilter, setProfessionFilter, setisCustomFilter } from "../actions/filters";

const Option = Select.Option;

class MainListFilter extends React.Component {

     handleToggleList = (e) => {
         this.props.handleToggleList(e.target.value);
     }

    capitalize = (string) => {
        return string.charAt(0).toUpperCase() + string.slice(1);
    }

    //DEBUG
    // componentDidMount() {
    //     console.log(this.props, 'props');
    // }
    // componentDidUpdate() {
    //     console.log(this.props, 'props');
    // }
    render() {
        return (
            <div className="InputTextFilter">
                <Radio.Group value={this.props.activeList} onChange={e => {this.props.dispatch(setisCustomFilter(e.target.value))}}>
                    <Radio.Button value={false}>Recipes</Radio.Button>
                    <Radio.Button value={true}>Reagents</Radio.Button>
                </Radio.Group>
                <label> Profession : </label>
                <Select defaultValue="all" style={{ width: 120 }} onChange={value => {this.props.dispatch(setProfessionFilter(value))}}>
                    <Option value="all">All</Option>
                    {
                        this.props.professions.map( profession => (
                            <Option value={profession.toLowerCase()}>{this.capitalize(profession)}</Option>
                        ))
                    }
                </Select>
                <input
                    type="text"
                    value={this.props.filters.text}
                    placeholder="Filter items"
                    onChange={e => {this.props.dispatch(setTextFilter(e.target.value))}} />
            </div>
        );
    }
}


const mapStateToProps = state => {
    console.log(state.recipes);
    let array = [];
    state.recipes.map( recipe => {recipe.professions.map( profession => {array.indexOf(profession) === -1 ? array.push(profession) : ''})})

    return {
        filters: state.filters,
        professions: array
    };
};

export default connect(mapStateToProps)(MainListFilter);