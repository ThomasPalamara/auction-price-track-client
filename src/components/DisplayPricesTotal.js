import React from "react";
import { connect } from 'react-redux';
import DisplayPricesTable from './DisplayPricesTable';
import PriceCoinDisplay from "./PriceCoinDisplay";
import { Row, Col } from 'antd';
import { editQuantityIngredient } from '../actions/ingredient';
import { editQuantityProduct } from '../actions/product';

class DisplayPricesTotal extends React.Component {
    constructor(props) {
        super(props);
        this.state = {};
    }

    getPriceForQuantity = (auctions, need) => {
        let qteactu = 0;
        let result = 0
        auctions.prices.map((element) => {
            if (need > qteactu) {
                let qte = Math.min(need - qteactu, element.quantity);
                result += qte * element.price
                qteactu += qte
                console.log(result);
            }
        });
        return result;
    }

    getTotal = () => {
        let totalProduct = 0;
        let totalIngredient = 0;

        this.props.product.map((element) => {
            const itemElem = this.props.auctionHouse.find(elem => elem.item_id === element.id.toString()); 
            totalProduct += this.getPriceForQuantity(itemElem.auctions[0], element.quantity);
        });

        this.props.ingredient.map((element) => {
            const itemElem = this.props.auctionHouse.find(elem => elem.item_id === element.id.toString()); 
            totalIngredient += this.getPriceForQuantity(itemElem.auctions[0], element.quantity);
        });

        const profit = totalProduct - totalIngredient;

        return {totalProduct, totalIngredient, profit};
    }

    render() {
        return (
            <Row>
                <Col span={8}><PriceCoinDisplay price={this.getTotal().totalProduct}/></Col>
                <Col span={8}><PriceCoinDisplay price={this.getTotal().profit}/></Col>
                <Col span={8}><PriceCoinDisplay price={this.getTotal().totalIngredient}/></Col>
            </Row>
        );
    }
}

const mapStatetoProps = state => {

    return {
        product: state.product,
        ingredient: state.ingredient

    }
};

export default connect(mapStatetoProps)(DisplayPricesTotal);