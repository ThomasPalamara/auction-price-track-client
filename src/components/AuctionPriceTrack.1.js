import React, { useState } from 'react';
import ReactSwipe from 'react-swipe';

import RealmSelection from './RealmSelection';
import MainList from './MainList';
import Stats from './Stats';

// TODO Ajouter la selection de language au store puis corrigÃ© celle en place et ajouter a display table

const AuctionPriceTrack = () => {
  // State par default pour faciliter les tests
  let reactSwipeEl;

  const [realm, setRealm] = useState({
    label: '',
    value: '',
  });

  const handleRealmPicked = (e) => {
    console.log(e, 'realm');
    setRealm(realm);
  };

  return (
    <div className="main">

      <button type="button" onClick={() => reactSwipeEl.prev()}>Previous</button>
      <button type="button" onClick={() => reactSwipeEl.next()}>Next</button>

      <ReactSwipe className="carousel" swipeOptions={{ continuous: false }} ref={(el) => { reactSwipeEl = el; }}>
        <RealmSelection handleRealmPicked={handleRealmPicked} />
        <MainList />
        <Stats realm={realm} />
      </ReactSwipe>

    </div>
  );
}

export default AuctionPriceTrack;
