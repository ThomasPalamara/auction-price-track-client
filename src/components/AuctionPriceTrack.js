import React from 'react';
import RealmSelection from './RealmSelection';
import MainList from './MainList';
import DisplayStats from './DisplayStats';
import { Divider } from 'antd';

// TODO Ajouter la selection de language au store puis corrigÃ© celle en place et ajouter a display table

export default class AuctionPriceTrack extends React.Component {
  state = {
    realm : 'Ysondre',
    auctionHouse: null,
    loading: false,
  };
  handleRealmPicked = (realm) => {
    this.setState({ realm });
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.realm && prevState.realm !== this.state.realm) {
      console.log('Fetching');
      this.setState({ loading: true });
      fetch(`/api/auctions/${this.state.realm}`)
        .then(res => res.json())
        .then(response => {
          this.setState({ loading: false });
          response.map((item) => {
            return item.auctions[0].prices.sort(function (a, b) { return (a.price > b.price) ? 1 : ((b.price > a.price) ? -1 : 0); });
          })
          return this.setState({ auctionHouse: response })
        }
      );
    }
  }

    render() {
    let displayPriceSection;
    if(!this.state.auctionHouse){
      displayPriceSection = <h3>Veuillez selectionner un serveur</h3>;
    }else{
      displayPriceSection = <DisplayStats auctionHouse={this.state.auctionHouse}/>
    }
    return (
      <div className="main">
        <RealmSelection loading={this.state.loading} handleRealmPicked={this.handleRealmPicked}/>
        <Divider/>
        <MainList/>
        <Divider/>
        {displayPriceSection}
      </div>
    );
  }
}
