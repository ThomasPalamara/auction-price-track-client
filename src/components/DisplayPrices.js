import React from "react";
import { connect } from 'react-redux';
import DisplayPricesTable from './DisplayPricesTable';
import DisplayPricesTotal from './DisplayPricesTotal';
import { Row, Col } from 'antd';
import { editQuantityIngredient } from '../actions/ingredient';
import { editQuantityProduct } from '../actions/product';


const DisplayPrices = (props) => (
    <div>
        <Row>
            <Col span={24}>
                <DisplayPricesTotal auctionHouse={props.auctionHouse} />
            </Col>
        </Row>
        <Row>
            <Col span={12}>
                <DisplayPricesTable auctionHouse={props.auctionHouse} editQuantity={editQuantityProduct} items={props.product} />
            </Col>
            <Col span={12}>
                <DisplayPricesTable auctionHouse={props.auctionHouse} editQuantity={editQuantityIngredient} items={props.ingredient} />
            </Col>
        </Row>
    </div>
);

const mapStatetoProps = state => {

    return {
        product: state.product,
        ingredient: state.ingredient

    }
};

export default connect(mapStatetoProps)(DisplayPrices);


//Tableau de toutes les auctions pour chaque item
// {this.props.product &&
//     this.props.product.map((item) => {
//         const itemElem = this.props.auctionHouse.find(elem => elem.item_id === item.id.toString())
//         const totalPrice = item.quantity
//         return (
//             <div>
//                 <p>{item.name}</p>
//                 <InputNumber min={0} defaultValue={1}
//                     onChange={(value) => {
//                         this.props.dispatch(editQuantityProduct({ id: item.id, quantity: value }))
//                     }} />
//             </div>
//         )
//     })}