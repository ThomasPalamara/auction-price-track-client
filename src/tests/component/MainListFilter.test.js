import React from 'react';
import { shallow } from "enzyme";
import toJSON from 'enzyme-to-json';
import { MainListFilter } from '../../components/MainListFilter';

test('Should render MainListFilter component correctly', () => {
    const filters = {
        text: '',
        profession: 'all',
        isCustom: false
    }
    const professions = ['enchant','alchemy']
    const wrapper = shallow(<MainListFilter professions={professions} filters={filters} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
})

// test('Should handle isCustom filter change correctly', () => {
//     const filters = {
//         text: '',
//         profession: 'all',
//         isCustom: false
//     }
//     const professions = ['enchant','alchemy']
//     const wrapper = shallow(<MainListFilter professions={professions} filters={filters} />);
//     wrapper.find('RadioGroup').simulate('change');
// })