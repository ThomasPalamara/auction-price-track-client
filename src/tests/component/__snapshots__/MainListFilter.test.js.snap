// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render MainListFilter component correctly 1`] = `
<div
  className="InputTextFilter"
>
  <RadioGroup
    buttonStyle="outline"
    disabled={false}
    onChange={[Function]}
    prefixCls="ant-radio"
  >
    <RadioButton
      prefixCls="ant-radio-button"
      value={false}
    >
      Recipes
    </RadioButton>
    <RadioButton
      prefixCls="ant-radio-button"
      value={true}
    >
      Reagents
    </RadioButton>
  </RadioGroup>
  <label>
     Profession : 
  </label>
  <Select
    choiceTransitionName="zoom"
    defaultValue="all"
    onChange={[Function]}
    prefixCls="ant-select"
    showSearch={false}
    style={
      Object {
        "width": 120,
      }
    }
    transitionName="slide-up"
  >
    <Option
      value="all"
    >
      All
    </Option>
    <Option
      key="0"
      value="enchant"
    >
      Enchant
    </Option>
    <Option
      key="1"
      value="alchemy"
    >
      Alchemy
    </Option>
  </Select>
  <input
    onChange={[Function]}
    placeholder="Filter items"
    type="text"
    value=""
  />
</div>
`;
